python/tests/integration/arcticdb/test_vector_db.py

```diff
@@ -0,0 +1,229 @@
+    with pytest.raises(ArcticUnsupportedDataTypeException):
+        vector_db["vdb"].upsert(True)
+    with pytest.raises(ArcticUnsupportedDataTypeException):
+        vector_db["vdb"].upsert([])
+
+
+def test_inextant_namespace(arctic_client):
+    ac = arctic_client
+    ac.create_library("pytest_test_inextant_namespace")
+    vector_db = VectorDB(ac["pytest_test_inextant_namespace"])
+    with pytest.raises(NoSuchVersionException):
+        vector_db.read("DOES_NOT_EXIST")
+    with pytest.raises(NoSuchVersionException):
+        vector_db["DOES_NOT_EXIST"].read()
+
+
+def test_vector_db_requires_library(arctic_client):
+    with pytest.raises(ArcticUnsupportedDataTypeException):
+        VectorDB(3)
+
+
+def test_vector_db_empty_upsertions(arctic_client):
+    np.random.seed(0)
+    random.seed(0)
+
+    ac = arctic_client
+    ac.create_library("pytest_test_vector_db_empty_upsertions")
+
+    vector_db = VectorDB(ac["pytest_test_vector_db_empty_upsertions"])
+
+    with pytest.raises(ValueError):
+        vector_db.upsert("empty1", np.array([[]]))
+
+    vector_db.upsert("empty2", {})
+    assert_frame_equal(vector_db.read("empty2"), pd.DataFrame(), check_index_type=False)
+
+    vector_db.upsert("empty3", pd.DataFrame())
+    assert_frame_equal(vector_db.read("empty3"), pd.DataFrame(), check_index_type=False)
+
+
+def test_vector_db_upsertion(arctic_client):
+    np.random.seed(0)
+    random.seed(0)
+
+    ac = arctic_client
+    ac.create_library("pytest_test_vector_db_upsertion")
+
+    vector_db = VectorDB(ac["pytest_test_vector_db_upsertion"])
+
+    dimensions, number_of_vectors, k = 15, 20, 5
+    first_string_column_names = [
+        "".join(random.choices(string.ascii_uppercase + string.digits, k=10)) for _ in range(number_of_vectors)
+    ]
+    second_string_column_names = [
+        "".join(random.choices(string.ascii_uppercase + string.digits, k=15)) for _ in range(number_of_vectors)
+    ]
+    first_df = pd.DataFrame(np.random.rand(dimensions, number_of_vectors), columns=first_string_column_names)
+    second_df = pd.DataFrame(np.random.rand(dimensions, number_of_vectors), columns=second_string_column_names)
+
+    vector_db.upsert("first_is_first", first_df)
+    vector_db.upsert("first_is_first", second_df)
+    vector_db.upsert("second_is_first", second_df)
+    vector_db.upsert("second_is_first", first_df)
+
+    assert_frame_equal(vector_db.read("first_is_first"), vector_db.read("second_is_first"), check_like=True)
+
+    vector_db.upsert("first_is_first", first_df)
+    vector_db.upsert("first_is_first", first_df)
+    vector_db.upsert("first_is_first", first_df)
+
+    assert_frame_equal(vector_db.read("first_is_first"), vector_db.read("second_is_first"), check_like=True)
+
+
+@use_of_function_scoped_fixtures_in_hypothesis_checked
+@given(df=data_frames([column("a", elements=numeric_type_strategies())], index=range_indexes()))
+def test_upsertion_idempotent(arctic_client, df):
```

Comment:
Needs fleshing out

Reply:
Hm I might just get rid of it since `test_vector_db_upsertion` tets idempotence, or maybe I'll use the same setup as the hypothesis test.

Reply:
(did the latter)